lib/chef/provider/powershell_script.rb

```diff
@@ -27,8 +27,71 @@ class PowershellScript < Chef::Provider::WindowsScript
       provides :powershell_script
 
       action :run do
-        validate_script_syntax!
-        super()
+        Chef::Log.debug("using inline impl: #{new_resource.use_inline_powershell}")
+        if new_resource.use_inline_powershell
+          # Because of the nature of powershell_exec, we can't easily
+          if new_resource.live_stream
+            raise "powershell_script does not support live_stream when inline powershell is enabled - please choose one or the other"
+          end
+
+          # This comes from the execute super resource - since we want to limit the scope of this to powershell,
+          # we use this here
+          if creates && sentinel_file.exist?
+            logger.debug("#{new_resource} sentinel file #{sentinel_file} exists - nothing to do")
+            return false
+          end
+
+          converge_by("execute direct powershell #{new_resource.name}") do
+            # For scoping - otherwise, we lose return_code/stdout/stderr when the timeout block is over
+            return_code = nil
+            stdout = nil
+            stderr = nil
+            Timeout.timeout(new_resource.timeout) do
+              r = powershell_exec!(inline_powershell_wrapper_script, new_resource.interpreter.to_sym)
+              # Split out the stdout/return code format if needed
+              return_code = r.result
```

Comment:
Seems like a lot of this code is already encapsulated by `chef/chef-powershell-shim:chef-powershell/lib/chef-powershell/powershell.rb` and friends... the timeout, the capturing of errors, etc.. (see [exec](https://github.com/chef/chef-powershell-shim/blob/main/chef-powershell/lib/chef-powershell/powershell.rb#L128))

Reply:
The honest answer to that I knew about powershell_exec, which doens't, but not Powershell.exec, which does the timeout and some of the demunging stuff.  I could absolutely move to that implentation, and use the timeout stuff there

Reply:
I'm fixing up my tests before pushing more code at this (working out how to make rspec acutally succesfully call FFI stuff... whee!), but I acutally do have a useful answer to this now - powershell_exec actually calls the stuff in chef-powershell, but also has the ability to switch interpreter on its own.  I _have_ refactored it to use the inbuilt timeout stuff, however, and am refactoring out some of the error handling complexirty

Reply:
Oh, to answer the actual question, the thing this does that just calling powershell_exec doesn't is be the powershell_script resource - since powershell_exec isn't a resource, to use it in an expected converge-time way, users either have to wrap it in ruby_block, have it in a unified mode resource, or wrap it in some other resource.  The point of this is to have a simple property that can be set to do things the performant but not-quite-as-featured way, rather than having all of that opacity.

This is particularly true if you want simple guards - you end up with messes like:

```
fast_powershell_thing "thing" do
   code "thing"
   not_if { powershell_exec!("guardthing") }
end
```

or similar monstrosities.  My actual initial implementation involved literally overriding only_if and not_if in the resource to make this read sanely, which does work in a custom resource, but that's a terrible, terrible pattern that I really want away from.

Reply:
> I'm fixing up my tests before pushing more code at this (working out how to make rspec acutally succesfully call FFI stuff... whee!), but I acutally do have a useful answer to this now - powershell_exec actually calls the stuff in chef-powershell, but also has the ability to switch interpreter on its own. I _have_ refactored it to use the inbuilt timeout stuff, however, and am refactoring out some of the error handling complexirty

@jaymzjulian are your test updates ready for review?

Reply:
> @jaymzjulian are your test updates ready for review?

they are